pkg/util/testing/json_roundtrip.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+type RoundTripTestCase struct {
+	Name   string
+	JSON   string
+	Object json.Marshaler
+
+	ExpectedMarshalError   string
+	ExpectedUnmarshalError string
+}
+
+// Performs JSON roundtrip testcase
+// Either JSON or Object may be provided, or both.
+// If only one is provided, the other will be inferred from unmarshalling/marshalling
+func DoJSONRoundTripTest(t *testing.T, exampleFactory func() json.Unmarshaler, testCases []RoundTripTestCase) {
```

Comment:
I really like this! For some reason I'm not a big fan of taking `testing.T` as a parameter, and I see your signature doesn't return an error.

I think I would try a signature like this:
```golang
func (t *RoundTripTestCase) RoundTripTest(exampleFactory func() json.Unmarshaler) error {
    // impl
}
```

which means the caller will probably do something like:
```golang
for name, test := range tests {
    t.Run(name, func(t *testing.T) {
        if err := test.RoundTripTest(unmarshaler); err != nil {
             t.Fatalf("test failed: %v", err)
        }
    })
}
```

Reply:
I refactored the RoundTripTest the way you mentioned :)